test/hotspot/gtest/runtime/test_mutex_rank.cpp

```diff
@@ -165,4 +195,36 @@ TEST_VM_ASSERT_MSG(MutexRank, monitor_wait_rank_special,
   monitor_rank_special_minus_one->unlock();
   monitor_rank_special->unlock();
 }
+
+TEST_VM_ASSERT_MSG(MutexRank, monitor_wait_access_leaf,
+                   ".* Attempting to acquire lock monitor_rank_leaf/11 out of order with lock monitor_rank_access/1 "
+                   "-- possible deadlock") {
+  JavaThread* THREAD = JavaThread::current();
+  ThreadInVMfromNative invm(THREAD);
+
+  Monitor* monitor_rank_access = new Monitor(Mutex::access, "monitor_rank_access", false, Mutex::_safepoint_check_never);
+  Monitor* monitor_rank_leaf = new Monitor(Mutex::leaf, "monitor_rank_leaf", false, Mutex::_safepoint_check_never);
+
+  monitor_rank_access->lock_without_safepoint_check();
+  monitor_rank_leaf->lock_without_safepoint_check();
+  monitor_rank_leaf->wait_without_safepoint_check(1);
```

Comment:
If you want to exercise the wait() error case you should lock the leaf lock first and then the access one. Then you will get the assert at wait_without_safepoint_check(1). You will have to fix the error message that we check for above to be the one used in the wait case.

Reply:
The intent is to lock `access` first, then `leaf` -- after all, that was the issue I initially found. About `wait`: I copy-pasted the shape of the test from the `monitor_wait_rank_special` above. I thought it was dubious that we basically expect to assert at `monitor_rank_leaf->lock_without_safepoint_check();`, but thought having `wait` makes the code a bit more readable. I can just remove `wait` here.

Reply:
Ah, you mean the test checks the attempt to wait! Right, let me fix that.

Reply:
Fixed. There is a wrinkle: `leaf` is not a `special` lock, so this test would make little sense. I replaced `leaf` with `tty` instead.

Reply:
Fix looks good. I thought you wanted to check the other failure mode in wait() (the one where you have to wait on the least rank owned monitor).